_posts/2018-10-12-kafka-stream-introduction.md

```diff
@@ -0,0 +1,257 @@
+        try {
+            Thread.sleep(250);
+        } catch (InterruptedException e) {}
+    });
+
+    producer.close();
+```
+
+This wil create 3 floors with a random device id, and afterwards send 10 times an event for a random door that it is opened.
+
+### Reading of the output
+
+For checking what happens in the system a data dumper was created that outputs all the messages on all the topics of interest (as well the input, as the output as the intermediate queues).
+
+```java
+public class DumpData {
+    private static Logger log = LoggerFactory.getLogger(DumpData.class);
+
+    public static void main(String... args) {
+        ExecutorService executorService = Executors.newFixedThreadPool(2);
+        executorService.submit(() -> {
+            KafkaConsumer<String, String> consumer = new KafkaConsumer<>(defaultProperties("your_client_id"));
+
+            List<String> topics = consumer.listTopics().keySet().stream()
+                    .filter(streamName -> streamName.startsWith("stream_"))
+                    .peek(log::info)
+                    .collect(Collectors.toList());
+
+            consumer.subscribe(topics);
+
+            while (true) {
+                ConsumerRecords<String, String> records = consumer.poll(100);
+                records.forEach(record -> {
+                            log.info("topic = {}, partition = {}, offset = {}, key = {}, value = {}",
+                                    record.topic(), record.partition(), record.offset(), record.key(), record.value());
+                        }
+                );
+            }
+        });
+    }
+}
+```
+This will subscribe to all the topics that start with 'stream_' on the Kafka. 
+
+### The main part
+
+So we finally arrived at the part where it all happens. 
+
+Just as a recap, the goal of this stream is to transform both input streams into a stream that gives how many people took the stairs at each floor. 
+
+As a start we must create a new StreamBuilder from the Kafka library
+```java
+    final StreamsBuilder builder = new StreamsBuilder();
+```
+
+We need to get the data somewhere, here we get it from the same topics our data simulater writes to.
+```java
+    KStream<String, String> sourceDev = builder.stream("stream_in_dev"); // which device is where
+    KStream<String, String> stream_in = builder.stream("stream_in"); // the pulse messages
+```
+
+The first stream (stream_in_dev) will be converted into a lookup table that is used when handling the second stream. 
+This lookup table will contain which device is installed in on what floor.
+```java
+    KTable<String, String> streamKtableDev = sourceDev
+        .groupByKey()
+        .reduce((val1, val2) -> val1, Materialized.as("stream_ktable_dev"));
+```
+Since one of the shortcuts we took is creating all the topics with only one partition we don't have any problems with streams not having the data it needs.
```

Comment:
```suggestion
Since one of the shortcuts we took is creating all the topics with only one partition, we don't have any problems with streams not having the data it needs.
```